File: server/python/src/metaobjects/codegen/template_codegen/output_pattern.py
"""Expands the cross-port output-pattern grammar (SP-1 §3.3): ``{name}``,
``{Name}`` (PascalCase), ``{package}`` (``::`` -> ``/``). An empty ``{package}``
collapses its trailing/leading slash so ``{package}/{name}`` with no package
yields just ``{name}``. Unknown placeholders raise.

Byte-equivalent to the TypeScript ``output-pattern.ts`` and the JVM
``OutputPattern``.
"""

from __future__ import annotations

import re

_TOKEN = re.compile(r"\{(\w+)\}")


def _pascal(s: str) -> str:
    return "".join(w[:1].upper() + w[1:] for w in re.split(r"[^A-Za-z0-9]+", s) if w)


def expand_output_pattern(
    pattern: str, name: str | None, package: str | None
) -> str:
    """Expand ``pattern``. ``name`` may be ``None`` (perPackage/perModel);
    ``package`` may be ``None``/empty."""
    pkg_empty = False

    def repl(m: re.Match[str]) -> str:
        nonlocal pkg_empty
        token = m.group(1)
        if token == "package":
            p = (package or "").replace("::", "/")
            if p == "":
                pkg_empty = True
            return p
        if token == "name":
            if name is None:
                raise ValueError(
                    f"output pattern {pattern!r} uses {{name}} but no entity name is in scope"
                )
            return name
        if token == "Name":
            if name is None:
                raise ValueError(
                    f"output pattern {pattern!r} uses {{Name}} but no entity name is in scope"
                )
            return _pascal(name)
        raise ValueError(f"unknown placeholder {{{token}}} in output pattern {pattern!r}")

    out = _TOKEN.sub(repl, pattern)
    if pkg_empty:
        out = re.sub(r"^/+", "", out)
        out = re.sub(r"/{2,}", "/", out)
    return out

File: server/java/codegen-base/src/main/java/com/metaobjects/generator/template/OutputPattern.java
package com.metaobjects.generator.template;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Expands the cross-port output-pattern grammar (SP-1 §3.3): {@code {name}},
 * {@code {Name}} (PascalCase), {@code {package}} ({@code ::} → {@code /}). An empty
 * {@code {package}} collapses its trailing/leading slash so {@code {package}/{name}}
 * with no package yields just {@code {name}}. Unknown placeholders are a hard error.
 *
 * <p>Byte-equivalent to the TypeScript {@code output-pattern.ts}.
 */
public final class OutputPattern {

    private OutputPattern() {}

    private static final Pattern TOKEN = Pattern.compile("\\{(\\w+)\\}");

    /**
     * @param pattern the output path pattern
     * @param name    the entity name (may be null for perPackage/perModel scopes)
     * @param pkg     the package (may be null/empty)
     */
    public static String expand(String pattern, String name, String pkg) {
        Matcher m = TOKEN.matcher(pattern);
        StringBuffer sb = new StringBuffer();
        boolean pkgEmpty = false;
        while (m.find()) {
            String token = m.group(1);
            String rep;
            switch (token) {
                case "package": {
                    String p = pkg == null ? "" : pkg.replace("::", "/");
                    if (p.isEmpty()) pkgEmpty = true;
                    rep = p;
                    break;
                }
                case "name":
                    if (name == null) {
                        throw new IllegalArgumentException(
                            "output pattern '" + pattern + "' uses {name} but no entity name is in scope");
                    }
                    rep = name;
                    break;
                case "Name":
                    if (name == null) {
                        throw new IllegalArgumentException(
                            "output pattern '" + pattern + "' uses {Name} but no entity name is in scope");
                    }
                    rep = pascal(name);
                    break;
                default:
                    throw new IllegalArgumentException(
                        "unknown placeholder {" + token + "} in output pattern '" + pattern + "'");
            }
            m.appendReplacement(sb, Matcher.quoteReplacement(rep));
        }
        m.appendTail(sb);
        String out = sb.toString();
        if (pkgEmpty) {
            out = out.replaceAll("^/+", "").replaceAll("/{2,}", "/");
        }
        return out;
    }

    private static String pascal(String s) {
        StringBuilder b = new StringBuilder();
        for (String w : s.split("[^A-Za-z0-9]+")) {
            if (w.isEmpty()) continue;
            b.append(Character.toUpperCase(w.charAt(0))).append(w.substring(1));
        }
        return b.toString();
    }
}
